File: BookVerse/src/types/book.js
// =====================================================
//  TYPE DEFINITIONS — Book Interface
//  Định nghĩa cấu trúc dữ liệu Book để đồng bộ với Java Backend
//  Trong JS thuần, đây là JSDoc để IDE (VSCode) hỗ trợ IntelliSense
// =====================================================

/**
 * @typedef {Object} BookCategory
 * @property {number} id
 * @property {string} name
 * @property {string} slug
 */

/**
 * @typedef {Object} Author
 * @property {number} id
 * @property {string} name
 * @property {string} [bio]
 * @property {string} [avatarUrl]
 */

/**
 * @typedef {Object} BookReview
 * @property {number} id
 * @property {number} rating       - Rating từ 1-5
 * @property {string} comment
 * @property {string} reviewerName
 * @property {string} createdAt
 */

/**
 * @typedef {'IN_STOCK' | 'OUT_OF_STOCK' | 'PRE_ORDER'} BookStatus
 * Khớp với enum Java: BookStatus.java
 */

/**
 * @typedef {'PAPERBACK' | 'HARDCOVER' | 'EBOOK' | 'SIGNED_EDITION'} BookFormat
 * Khớp với enum Java: BookFormat.java
 */

/**
 * @typedef {Object} Book
 * @property {number}       id
 * @property {string}       title
 * @property {string}       slug          - URL-friendly unique identifier
 * @property {Author}       author
 * @property {BookCategory} category
 * @property {string}       description
 * @property {number}       price
 * @property {number}       [originalPrice] - Giá gốc (nếu đang giảm)
 * @property {number}       [discountPercent]
 * @property {string}       coverImageUrl
 * @property {string[]}     [galleryImages]
 * @property {BookFormat}   format
 * @property {BookStatus}   status
 * @property {number}       pageCount
 * @property {string}       isbn
 * @property {string}       publisher
 * @property {string}       publishedDate  - ISO 8601 format
 * @property {string}       language
 * @property {number}       rating         - Average rating (1-5)
 * @property {number}       reviewCount
 * @property {number}       soldCount
 * @property {boolean}      isFeatured
 * @property {boolean}      isEditorChoice - "Editor's Choice" badge
 * @property {string}       createdAt
 * @property {string}       updatedAt
 */

/**
 * @typedef {Object} PaginatedResponse
 * @property {Book[]} content
 * @property {number} totalElements
 * @property {number} totalPages
 * @property {number} currentPage
 * @property {number} pageSize
 * @property {boolean} hasNext
 * @property {boolean} hasPrevious
 */

/**
 * @typedef {Object} BooksState
 * @property {Book[]}            bestsellers
 * @property {Book[]}            featured
 * @property {Book[]}            newArrivals
 * @property {Book|null}         selectedBook
 * @property {PaginatedResponse|null} searchResults
 * @property {'idle'|'loading'|'succeeded'|'failed'} status
 * @property {string|null}       error
 */

// Xuất constants cho giá trị enum — dùng trong toàn dự án
export const BOOK_STATUS = {
  IN_STOCK:    'IN_STOCK',
  OUT_OF_STOCK:'OUT_OF_STOCK',
  PRE_ORDER:   'PRE_ORDER',
};

export const BOOK_FORMAT = {
  PAPERBACK:      'PAPERBACK',
  HARDCOVER:      'HARDCOVER',
  EBOOK:          'EBOOK',
  SIGNED_EDITION: 'SIGNED_EDITION',
};

File: BookVerse-Backend/src/main/java/com/bookverse/entity/Book.java
package com.bookverse.entity;

import jakarta.persistence.*;
import lombok.*;

import java.math.BigDecimal;
import java.time.LocalDate;

@Entity
@Table(name = "books")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Book {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    @Column(nullable = false, unique = true)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(nullable = false)
    private BigDecimal price;

    @Column(name = "original_price")
    private BigDecimal originalPrice;

    @Column(name = "discount_percent")
    private Integer discountPercent;

    @Column(name = "cover_image_url")
    private String coverImageUrl;

    @ManyToOne
    @JoinColumn(name = "category_id")
    private Category category;

    private String author;

    private Double rating;

    @Column(name = "review_count")
    private Integer reviewCount;

    @Column(name = "is_editor_choice")
    private Boolean isEditorChoice;

    @Column(name = "is_featured")
    private Boolean isFeatured;

    @Column(name = "is_bestseller")
    private Boolean isBestseller;

    @Column(name = "is_new_arrival")
    private Boolean isNewArrival;

    private String isbn;

    private String publisher;

    @Column(name = "page_count")
    private Integer pageCount;

    @Column(name = "published_date")
    private LocalDate publishedDate;

    private String language;

    @Column(name = "stock_quantity")
    private Integer stockQuantity;
}
